R/check_structure.R

```diff
@@ -0,0 +1,135 @@
+#' @title Checks protocol document structure
+#'
+#' @description This function reads the protocol and checks if the document
+#' structure is correct: chunks have head and tail, required headings are
+#' present and in the right order,...
+#' This function is intended for checking if a protocol is ready to be rendered
+#' and published.
+#'
+#' @inheritParams get_path_to_protocol
+#'
+#' @return If one of the checks fails, an error message will be returned.
```

Comment:
A `checklist` function creates a [vector of problems](https://github.com/inbo/checklist/blob/84cd5114dbc8125c6c99721716fca16e5a92375a/R/check_filename.R#L158) and [adds that vector](https://github.com/inbo/checklist/blob/84cd5114dbc8125c6c99721716fca16e5a92375a/R/check_filename.R#L171) to the object. The trick I'm using is `"problem description"[condition]`. If condition is `FALSE`, it returns a vector of length 0. Otherwise a vector of length 1. I use this to reduce the number of if statements.

The object determines if the checks [failed](https://github.com/inbo/checklist/blob/84cd5114dbc8125c6c99721716fca16e5a92375a/R/checklist_class.R#L203) either because you didn't run all required tests or some of test failed. I use that status at the end of the [high level functions](https://github.com/inbo/checklist/blob/df9f02b08456cde6f90c7cd2af13a93c8e4e94b7/R/check_package.R#L85). Note that the high level functions have a `fail` arguments that determines how the function should react to failure. `TRUE` returns an error upon failure while `FALSE` only displays the message.

Reply:
The `checklist` way of doing things is nice to the user because you can run all checks and report all problems at once. At https://github.com/inbo/indicatoren we stop with an error as soon as an error occurs. This requires the user to rerun the tests several times.

Reply:
As I can't reply to the below comment, I do it here (I suppose the above explanation of @ThierryO is also on the below comment?).

- `checklist::check_source()`: if you prefer to keep `check_frontmatter()` and `check_structure()` as 2 separate functions, that's fine as well.  I can run them one after the other in protocolsource, it doesn't need a wrapper.
- in any case I suppose you'll have to change `assert_that()` with an object that collects the warnings and errors and finally presents them in a report.

Reply:
>The trick I'm using is `"problem description"[condition]`. If condition is `FALSE`, it returns a vector of length 0. Otherwise a vector of length 1. I use this to reduce the number of if statements.

I see, so we need to rewrite our `assert_that()` as `"problem description"[condition]`

> The object determines if the checks [failed](https://github.com/inbo/checklist/blob/84cd5114dbc8125c6c99721716fca16e5a92375a/R/checklist_class.R#L203) either because you didn't run all required tests or some of test failed. 

Got it. The R6class would need to be rewritten for the protocolhelper package unless we decide to enforce the use of `checklist::check_source()` (which would imply a lot of work)

> I use that status at the end of the [high level functions](https://github.com/inbo/checklist/blob/df9f02b08456cde6f90c7cd2af13a93c8e4e94b7/R/check_package.R#L85). Note that the high level functions have a `fail` arguments that determines how the function should react to failure. `TRUE` returns an error upon failure while `FALSE` only displays the message.

Yes, I saw that and it is very useful in GHA to set this to TRUE and in interactive mode to FALSE.

> * `checklist::check_source()`: if you prefer to keep `check_frontmatter()` and `check_structure()` as 2 separate functions, that's fine as well.  I can run them one after the other in protocolsource, it doesn't need a wrapper.

I think (not sure) a wrapper (= the high level functions which Thierry mentions) reduces the amount of code needed because than only the wrapper needs to have the fail or not functionality. But I agree, both approaches would work. I have no strong preference.

> * in any case I suppose you'll have to change `assert_that()` with an object that collects the warnings and errors and finally presents them in a report.

Yes, see how Thierry did this with `"problem description"[condition]` and collect these in a vector.

Reply:
> Got it. The R6class would need to be rewritten for the protocolhelper package unless we decide to enforce the use of `checklist::check_source()` (which would imply a lot of work)

Create a dedicated class for `protocolhelper` to handle only the specific (extra) checks. And use `checklist::check_source()` for the generic tests. It is OK to run multiple GHA.